routing/payment_lifecycle.go

```diff
@@ -390,6 +294,136 @@ func (p *shardHandler) launchShard(rt *route.Route) (*channeldb.HTLCAttemptInfo,
 }
 
+// shardResult holds the resulting outcome of a shard sent.
+type shardResult struct {
+	// preimage is the payment preimage in case of a settled HTLC. Only set
+	// if err is non-nil.
+	preimage lntypes.Preimage
+
+	// err indicates that the shard failed.
+	err error
+}
+
+// collectResult waits for the result for the given attempt to be available
+// from the Switch, then records the attempt outcome with the control tower. A
+// shardResult is returned, indicating the final outcome of this HTLC attempt.
+func (p *shardHandler) collectResult(attempt *channeldb.HTLCAttemptInfo) (
+	*shardResult, error) {
+
+	// Regenerate the circuit for this attempt.
+	_, circuit, err := generateSphinxPacket(
+		&attempt.Route, p.paymentHash[:],
+		attempt.SessionKey,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Using the created circuit, initialize the error decrypter so we can
+	// parse+decode any failures incurred by this payment within the
+	// switch.
+	errorDecryptor := &htlcswitch.SphinxErrorDecrypter{
+		OnionErrorDecrypter: sphinx.NewOnionErrorDecrypter(circuit),
+	}
+
+	// Now ask the switch to return the result of the payment when
+	// available.
+	resultChan, err := p.router.cfg.Payer.GetPaymentResult(
+		attempt.AttemptID, p.paymentHash, errorDecryptor,
+	)
+	switch {
+
+	// If this attempt ID is unknown to the Switch, it means it was never
+	// checkpointed and forwarded by the switch before a restart. In this
+	// case we can safely send a new payment attempt, and wait for its
+	// result to be available.
+	case err == htlcswitch.ErrPaymentIDNotFound:
+		log.Debugf("Payment ID %v for hash %x not found in "+
+			"the Switch, retrying.", attempt.AttemptID,
+			p.paymentHash)
+
+		cErr := p.failAttempt(attempt, err)
+		if cErr != nil {
+			return nil, cErr
+		}
+
+		return &shardResult{
+			err: err,
+		}, nil
+
+	// A critical, unexpected error was encountered.
+	case err != nil:
+		log.Errorf("Failed getting result for attemptID %d "+
+			"from switch: %v", attempt.AttemptID, err)
+
+		return nil, err
+	}
+
+	// The switch knows about this payment, we'll wait for a result to be
+	// available.
+	var (
+		result *htlcswitch.PaymentResult
+		ok     bool
+	)
+
+	select {
+	case result, ok = <-resultChan:
+		if !ok {
+			return nil, htlcswitch.ErrSwitchExiting
+		}
+
+	case <-p.router.quit:
+		return nil, ErrRouterShuttingDown
+	}
+
+	// In case of a payment failure, fail the attempt with the control
+	// tower and return.
+	if result.Error != nil {
+		log.Errorf("Attempt to send payment %x failed: %v",
+			p.paymentHash, result.Error)
+
+		err := p.failAttempt(attempt, result.Error)
+		if err != nil {
+			return nil, err
+		}
+
+		return &shardResult{
+			err: result.Error,
+		}, nil
+	}
+
+	// We successfully got a payment result back from the switch.
+	log.Debugf("Payment %x succeeded with pid=%v",
+		p.paymentHash, attempt.AttemptID)
+
+	// Report success to mission control.
+	err = p.router.cfg.MissionControl.ReportPaymentSuccess(
```

Comment:
That is still valid I think, since even though it fails in the next step, MissionControl should be given the information that the attempted route succeeded.